### PR TITLE
Implemented LTE, GTE opcode

### LINKED ISSUE #229: VM: implement LTE, GTE opcode

### PR DESCRIPTION
**Pull request by dauTT**
Closes #229

### CODE REVIEW

File: pkg/vm/stack/Int.go [hunk 1]
```diff
@@ -100,6 +100,22 @@ func (i *Int) Value() *big.Int {
 	return i.val
 }
 
+// Lte returns a bool value from the comparison of two integers, a and b.
+// value is true if a <= b.
+// value is false if a > b.
+func (i *Int) Lte(s *Int) bool {
+	res := i.Value().Cmp(s.Value())
```
Comment (fyrchik) [inline pkg/vm/stack/Int.go]: maybe `i.Value().Cmp(s.Value()) != -1` ?

Comment (dauTT) [reply pkg/vm/stack/Int.go]: :+1:

File: pkg/vm/vm_ops.go [hunk 2]
```diff
@@ -8,7 +8,9 @@ var opFunc = map[stack.Instruction]stackInfo{
 	stack.BOOLAND:     BoolAnd,
 	stack.BOOLOR:      BoolOr,
 	stack.LT:          Lt,
+  stack.LTE:         Lte,
```
[the remark this answers is not in the document]
Comment (dauTT) [reply pkg/vm/vm_ops.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go [hunk 3]
```diff
@@ -327,6 +327,64 @@ func TestNegateOp(t *testing.T) {
 	assert.Equal(t, int64(20), item.Value().Int64())
 }
 
+func TestLteOp(t *testing.T) {
+
+	v := VM{}
+
+	a, err := stack.NewInt(big.NewInt(10))
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (dauTT) [reply pkg/vm/vm_ops_maths_test.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go [hunk 4]
```diff
@@ -327,6 +327,64 @@ func TestNegateOp(t *testing.T) {
...
+
+	b, err := stack.NewInt(big.NewInt(10))
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (dauTT) [reply pkg/vm/vm_ops_maths_test.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go [hunk 5]
```diff
@@ -327,6 +327,64 @@ func TestNegateOp(t *testing.T) {
...
+
+	assert.Equal(t, true, item.Value())
+}
+
+func TestGteOp(t *testing.T) {
+
+	v := VM{}
+
+	a, err := stack.NewInt(big.NewInt(10))
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (dauTT) [reply pkg/vm/vm_ops_maths_test.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go [hunk 6]
```diff
@@ -327,6 +327,64 @@ func TestNegateOp(t *testing.T) {
...
+
+	b, err := stack.NewInt(big.NewInt(2))
+	assert.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (dauTT) [reply pkg/vm/vm_ops_maths_test.go]: :+1:

File: pkg/vm/vm_ops_maths_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (im-kulikov) [reply pkg/vm/vm_ops_maths_test.go]: it's bad practice.. must be used `assert.NoError` and `assert.Error`

File: pkg/vm/vm_ops_maths_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (im-kulikov) [reply pkg/vm/vm_ops_maths_test.go]: it's bad practice.. must be used `assert.NoError` and `assert.Error`

File: pkg/vm/vm_ops_maths_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (im-kulikov) [reply pkg/vm/vm_ops_maths_test.go]: it's bad practice.. must be used `assert.NoError` and `assert.Error`

File: pkg/vm/vm_ops_maths_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (im-kulikov) [reply pkg/vm/vm_ops_maths_test.go]: it's bad practice.. must be used `assert.NoError` and `assert.Error`

**Comment by dauTT** [comment]
@im-kulikov / @decentralisedkev ,

I actually can't see the difference between:

```
assert.Nil(t, err)
assert.NoError(t, err)
```

Is there an example where the application of the two asserts results in a truly discrepancy?

**Comment by decentralisedkev** [comment]
@im-kulikov @fyrchik @dauTT 

Thanks for reviewing the PR.

From the testify repo: https://github.com/stretchr/testify#assert-package

It says that assert.Nil(t, object) is good for errors. I checked the source code and given an error, both functions do the same thing. I decided to use assert.Nil(t, err) because we can also use the same method to check for nil objects and not just errors. Whereas with assert.NoError, we are limited by just the error interface.

If you still believe that this should not be done, then please open up an issue stating what the bad practice is. I believe my comment about assert.Nil for consistency across the codebase still stands.

**Comment by decentralisedkev** [comment]
> Or is just a matter of convention?

It's mostly a matter of taste.

**Comment by im-kulikov** [comment]
https://github.com/stretchr/testify/blob/master/assert/assertions.go#L1183

> I believe my comment about assert.Nil for consistency across the codebase still stands.

What consistency are you talking about? Nil, says nothing about the brevity of the absence of errors. At the same time, Error, NoError and EqualError clearly say that we are checking for the presence, absence or clear correspondence of an error. I don't see any connection to the consistency of the codebase.

Open the repository of famous companies that use go and assert, try to make them PR, on an attempt to replace NoError with Nil, you will be asked to return back. That's what I'm talking about.

**Comment by decentralisedkev** [comment]
> Open the repository of famous companies that use go and assert, try to make them PR, on an >attempt to replace NoError with Nil, you will be asked to return back. That's what I'm talking about.

The codebase uses assert.Nil throughout, so I wanted to keep the consistency, whenever possible.

In terms of famous companies, I'm not sure they would mind to be completely honest.

For example:

Docker uses both:

https://github.com/search?q=org%3Adocker+%22assert.Nil%28t%2C+err%29%22&type=Code
https://github.com/search?q=org%3Adocker+%22assert.NilError%28t%2C+err%29%22&type=Code

Dropbox and Turnitin too:

https://github.com/dropbox/changes-client/blob/2af50345ba4844881f4f6f9be82051ac0e935933/adapter/lxc/container_test.go#L33

https://github.com/search?q=org%3Aturnitin+%22require.Nil%28t%2Cerr%29%22&type=Code

Uber uses a different framework, but translated to testify, it becomes `assert.Nil(t, err)` :

https://github.com/uber/aresdb/blob/master/memstore/cuckoo_index_test.go#L38

Personally, I would have liked to use just `assert.Nil` which is what my initial comment was referring to when I said consistency.

If you want to change it, then I think opening a separate issue would be better.

**Comment by im-kulikov** [comment]
As I understand, this is your personal preference and there is no explanation about the consistency of the code.

**Comment by hal0x2328** [comment]
100 points dauTT

**Comment by lllwvlvwlll** [comment]
https://neoscan.io/transaction/630c74684cc6dad330397c63e8c7aa6403c043cc1f241638cff87e3b97a40716

**Pull request merged**